security/oc_certs.c

```diff
@@ -523,21 +452,88 @@ oc_certs_extract_first_role(const mbedtls_x509_crt *cert, oc_string_t *role,
        * (CN) component and store it.
        */
       dnAuthority = oc_certs_CN_extract_issuer(cert);
+      if (dnAuthority == NULL) {
+        return false;
+      }
     }
 
-    if (dnAuthority == NULL) {
+    // both role and authority are set
+    oc_new_string(role, (const char *)dnRole->val.p, dnRole->val.len);
+    oc_new_string(authority, (const char *)dnAuthority->val.p,
+                  dnAuthority->val.len);
+    return true;
+  }
+  return false;
+}
+
+#else /* MBEDTLS_VERSION_NUMBER > 0x03010000 */
+
+static bool
+certs_extract_first_role(const mbedtls_x509_crt *cert, oc_string_t *role,
+                         oc_string_t *authority)
+{
+  for (const mbedtls_x509_sequence *names = &cert->subject_alt_names;
+       names != NULL; names = names->next) {
+
+    mbedtls_x509_subject_alternative_name san;
+    int ret = mbedtls_x509_parse_subject_alt_name(&names->buf, &san);
+    if (ret < 0) {
+      return false;
+    }
+
+    if (san.type != MBEDTLS_X509_SAN_DIRECTORY_NAME) {
+      mbedtls_x509_free_subject_alt_name(&san);
+      continue;
+    }
+
+    const mbedtls_x509_name *dnRole = NULL;
+    const mbedtls_x509_name *dnAuthority = NULL;
+    for (const mbedtls_x509_name *name = &san.san.directory_name;
+         name != NULL && (dnRole == NULL || dnAuthority == NULL);
+         name = name->next) {
+      /* Look for the Common Name (CN) component in the directoryName */
+      if (oc_certs_DN_is_CN(name)) {
+        dnRole = name;
+        continue;
+      }
+      /* Look for an Organizational Unit (OU) component in the directoryName
+       */
+      if (oc_certs_DN_is_OU(name)) {
+        dnAuthority = name;
+        continue;
+      }
+    }
+
+    if (dnRole == NULL) {
+      mbedtls_x509_free_subject_alt_name(&san);
       return false;
     }
 
+    if (dnAuthority == NULL) {
+      /* If the OU component was absent in the directoryName, it is assumed
+       * that the issuer of this role certificate is the "authority".
+       * Accordingly, extract the issuer's name from the issuer's Common Name
+       * (CN) component and store it.
+       */
+      dnAuthority = oc_certs_CN_extract_issuer(cert);
+      if (dnAuthority == NULL) {
+        mbedtls_x509_free_subject_alt_name(&san);
+        return false;
```

Comment:
I this something new ? Could you link a OCF spec ?

Reply:
both the comment and code are old, I just kept it, hopefully it's mentioned somewhere in the OCF spec